### ISSUE TITLE
Error on compile

### ISSUE THREAD

**Issue by goschtl** [pauleveritt/htm_transcrypt#1]
Hi Paul,

i tried your example. I got a Problem when i transpile it. It seems the transscript does not have the right environment to get everything working.

Any idea whats happening here?

Transcrypt (TM) Python to JavaScript Small Sane Subset Transpiler Version 3.7.16
Copyright (C) Geatec Engineering. License: Apache 2.0

Saving target code in: /Users/ck/work/htm/htm_transcrypt/__target__/org.transcrypt.__runtime__.js
Traceback (most recent call last):
  File "/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 2923, in revisit_ImportFrom
    module = self.useModule ('{}.{}'.format (node.module, alias.name))  # So, attempt to use alias as a module
  File "/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 922, in useModule
    return self.module.program.provide (name, filter = self.filterId)   # Must be done first because it can generate a healthy exception
  File "/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 120, in provide
    return Module (self, moduleName, __moduleName__, filter)
  File "/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 129, in __init__
    self.findPaths (filter)
  File "/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 314, in findPaths
    message = '\n\tImport error, can\'t find any of:\n\t\t{}\n'.format ('\n\t\t'. join (self.program.searchedModulePaths))
org.transcrypt.utils.Error: Error while compiling (offending file last):
	File '/Users/ck/work/htm/htm_transcrypt/main.py', line 3, at import of:
	File 'myhtm', line 0, namely:

	Import error, can't find any of:
		/Users/ck/work/htm/htm_transcrypt/myhtm/htm.py
		/Users/ck/work/htm/htm_transcrypt/myhtm/htm.js
		/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/myhtm/htm.py
		/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/myhtm/htm.js
		/Users/ck/work/htm/htm_transcrypt/myhtm/htm.py
		/Users/ck/work/htm/htm_transcrypt/myhtm/htm.js
		/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/myhtm/htm.py
		/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/myhtm/htm.js
		/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/lib-dynload/myhtm/htm.py
		/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/lib-dynload/myhtm/htm.js
		/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/myhtm/htm.py
		/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/myhtm/htm.js
		/Users/ck/work/htm/htm_transcrypt/htm.py/myhtm/htm.py
		/Users/ck/work/htm/htm_transcrypt/htm.py/myhtm/htm.js

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/org/transcrypt/compiler.py", line 1475, in include
    node.args [0], '\n\t\t'. join (searchedIncludePaths)
org.transcrypt.utils.Error: Error while compiling (offending file last):
	File '/Users/ck/work/htm/htm_transcrypt/main.py', line 3, at import of:
	File 'myhtm', line 4, namely:

	Attempt to include file: <_ast.Str object at 0x1109cfc50>
	Can't find any of:
		/Users/ck/work/htm/htm_transcrypt/node_modules/htm/dist/htm.js
		/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/transcrypt/modules/node_modules/htm/dist/htm.js
		/Users/ck/work/htm/htm_transcrypt/node_modules/htm/dist/htm.js
		/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/node_modules/htm/dist/htm.js
		/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/lib-dynload/node_modules/htm/dist/htm.js
		/Users/ck/work/htm/htm_transcrypt/lib/python3.7/site-packages/node_modules/htm/dist/htm.js
		/Users/ck/work/htm/htm_transcrypt/htm.py/node_modules/htm/dist/htm.js

Saving target code in: /Users/ck/work/htm/htm_transcrypt/__target__/myhtm.js
Saving target code in: /Users/ck/work/htm/htm_transcrypt/__target__/main.js

Ready

Thanks in Advance
Christian

**Comment by pauleveritt** [comment]
It is very much an experiment from @jviide so I'm not able (nor is he) to really dig into it at the moment. Sorry.

**Comment by pauleveritt** [comment]
I should add though, thanks for looking at this, it would be great if we could make some progress.

**Comment by goschtl** [comment]
Yes, i will come back to it. I am just playing with your viewdom project.

**Comment by pauleveritt** [comment]
Great, thanks. As FYI, the next layers are:

- [viewdom_wired](https://viewdom-wired.readthedocs.io/en/latest/)
- Themester, a collection of services for making UIs that work between systems (e.g. Sphinx, Pelican, Flask)
- The important one: a Sphinx theme
- Last, the important one to me: a competitive SSG

I've done all of these in previous incarnations, just putting the pieces back together again.

**Comment by goschtl** [comment]
Hey Paul,

maybe a stupid question but in which kind of framework do you want run your viewdom_wired stuff. Do you have something special in mind for this? I am a long time zope/zope3 user and i know all of this concepts. It would be nice to have some easy/light and flexible traversal thing to get the models/contexts for the views. Do you have thought on that? Or can you recommend something...

Christian

**Comment by pauleveritt** [comment]
@goschtl htm has "bindings" such as Preact. I hope ultimately to have a binding to Pyramid and maybe to Flask. My primary interest is SSGs though, with my first interest in Sphinx. I had a "knowledge base for Sphinx" project almost done 1.5 years ago, took a detour.

I am currently doing resource trees, resource-oriented view lookups, and the like in a layer atop viewdom_wired. On one hand, my experience with Pyramid has taught me that it is tough to convince people. OTOH, if one did something really compelling for Sphinx, it would get adopted only for use in that.

I'd love advice from you or even participation.

**Comment by goschtl** [comment]
given you have this tiny little components like

Menu --> MenuItem; Header, Footer elements like this. These are shareable as it is just python.
What would you recommend to apply a CSS-Style to it. Do you see this totally out of scope, or do you have something in mind to add styles to it. I think it can be easy possible to create f.e. bootstrap elements out of it. It is possible to use it in different projects.

<Form> <Field> etc..

**Comment by pauleveritt** [comment]
I'll leave two comments to your post. You're right that, if I could get Transcrypt to treat my `viewdom_wired` dataclass component as a little DSL, then I can get client-side rehydration and re-rendering. Which is pretty neat.

I wanted add, though, that even if we couldn't get the Python->JS working, we could still serialize the VDOM. We could then do partial updates in the browser with Preact or something smaller, and use something like IntercoolerJS or TurboLinks (though those work with HTML fragments.)

**Comment by pauleveritt** [comment]
On to your second point, you're right that a `styled-components` kind of thing might be interesting. Not necessarily to do dynamic CSS client-side, but to ease SSR of CSS and be nicely encapsulated in the components. Haven't thought too much about that yet.

**Comment by goschtl** [comment]
Hey Paul, me again...

I think i have now an a better plan. I am not 100% convinced that python2js on this is that good idea. I will try to do the following: I will made a simple html-page which preact or something like that. To feed the preact.render i try to establish a Web-Socket Connection to Starlette or something and return the htm Syntax. I will try then to intercept all JS events and move then to the  Web-Socket Connection return again htm Syntax and rerender the result in the DOM. 

I only need some time now between kids projects corona etc.

P.S. i think it should be easy possible then to mix  client side htm and server side htm/jsx.
I try to fiddle around a bit and let you know where it ends...

Cheers and thanks for your time.
Christian

Btw. i saw lately this thread which goes into the same direction.
https://twitter.com/dan_abramov/status/1259614150386425858

**Comment by pauleveritt** [comment]
I'm still interested in transpiling, but mostly in support of the scenario you describe. I'd like an SSG that can do a TurboLinks/IntercoolerJS style of client-side page loading, possibly of VDOM fragments pumped into Preact rather than HTML fragments. I'm also interested in offline-first and prefetch to lessen the need for web sockets.

I'd still like things that really are client-side to have a small Python component converted to JS. But I view that as progressive enhancement.
